options.go

```diff
@@ -249,19 +248,14 @@ func (o RetryOptions) WithCondition( // nolint:gocritic
 	return o
 }
 
-func (o RetryOptions) ruleForException(err Error) RetryRule { // nolint:gocritic,lll
-	var edbErr Error
-	if !errors.As(err, &edbErr) {
-		panic(fmt.Sprintf("unexpected error type: %T", err))
-	}
-
+func (o RetryOptions) ruleForException(err Error) (RetryRule, error) {
 	switch {
-	case edbErr.Category(TransactionConflictError):
-		return o.txConflict
-	case edbErr.Category(ClientError):
-		return o.network
+	case err.Category(TransactionConflictError):
+		return o.txConflict, nil
+	case err.Category(ClientError):
+		return o.network, nil
 	default:
-		panic(fmt.Sprintf("unexpected error type: %T", err))
+		return RetryRule{}, fmt.Errorf("unexpected error type: %T", err)
```

Comment:
I don't think this is written down anywhere (it probably should be): all errors generated in the top level package should be one of the edgedb error types defined in `generatederrors.go`, and all errors returned from `internal/` packages should be wrapped in one of these. It is also not documented what each of the error types means or when it should be used partly because there are plans to rework the error system. I think this should be `&clientError{...}`.